src/common/commonutils/CommandUtils.c

```diff
@@ -29,270 +31,200 @@ static int NormalizeStatus(int status)
+    if (0 == workerPid)
+    {
+        // Child process
+        close(pipefd[0]);
+        if (STDOUT_FILENO != dup2(pipefd[1], STDOUT_FILENO))
+        {
+            exit(-1);
+        }
+        if (STDERR_FILENO != dup2(pipefd[1], STDERR_FILENO))
+        {
+            exit(-1);
+        }
+        execl("/bin/sh", "sh", "-c", command, (char*)NULL);
+        // If execl() fails, exit with the error code
+        exit(errno);
+    }
+    else
+    {
+        // Main process
+        const int defaultCommandTimeout = 60;  // seconds
+        const int callbackIntervalSeconds = 5; // seconds
+        int status = -1;
+        int childStatus = 0;
+        int bytesRead = 0;
+        int inputBufferPos = 0;
+        unsigned int outputBufferPos = 0;
+        unsigned int outputBufferSize = 0;
+        time_t startTime = time(NULL);
+        time_t lastCallbackTime = 0;
+        time_t currentTime = 0;
+        char buffer[BUFFER_SIZE];
+
+        close(pipefd[1]);
+
+        if ((NULL != callback) && (timeoutSeconds == 0))
+        {
+            timeoutSeconds = defaultCommandTimeout;
+        }
 
-        if (0 == (intermediateProcess = fork()))
+        for (;;)
         {
-            // Intermediate process
+            fd_set fdset;
+            int ret = 0;
+            char* tmp = NULL;
 
-            if (0 == (workerProcess = fork()))
-            {
-                // Worker process
-                status = execl("/bin/sh", "sh", "-c", command, (char*)NULL);
-                _exit(status);
-            }
-            else if (workerProcess < 0)
-            {
-                // Intermediate process
-                OsConfigLogError(log, "Failed forking process to execute command (errno: %d, '%s')", errno, strerror(errno));
-                status = -1;
+            FD_ZERO(&fdset);
+            FD_SET(pipefd[0], &fdset);
 
-                // Kill the timer process if created and wait on it before exiting otherwise timer becomes a zombie process
-                if (timerProcess > 0)
-                {
-                    KillProcess(timerProcess);
-                    waitpid(timerProcess, &intermediateStatus, 0);
+            ret = select(pipefd[0] + 1, &fdset, NULL, NULL, &(struct timeval){.tv_sec = 0, .tv_usec = 500 * 1000});
+            if (ret < 0)
+            {
+                if (EINTR == errno) {
+                    continue;
                 }
-                _exit(status);
+                OsConfigLogError(log, "Error doing select for command '%s' - %s", command, strerror(errno));
+                status = -1;
+                break;
             }
 
-            if (0 == (timerProcess = fork()))
+            currentTime = time(NULL);
+            if ((timeoutSeconds > 0) && (currentTime - startTime >= timeoutSeconds))
             {
-                // Timer process
+                OsConfigLogError(log, "Timeout reading from pipe for command '%s'", command);
```

Comment:
>"Timeout reading from pipe for command '%s'", command); [](http://example.com/codeflow?start=38&length=55)

Logging should tell the user when the timeout for the command execution is exceeded, not just reading from pipe. Do we have this?